Index: object.di
===================================================================
--- object.di	(revision 2167)
+++ object.di	(working copy)
@@ -47,13 +47,10 @@
     // 8:       // has constructors
     void*       deallocator;
     OffsetTypeInfo[] offTi;
-  version( DigitalMars )
-  {
     void*       defaultConstructor;

     static ClassInfo find(char[] classname);
     Object create();
-  }
 }

 struct OffsetTypeInfo
Index: lib/common/tango/core/Thread.d
===================================================================
--- lib/common/tango/core/Thread.d	(revision 2167)
+++ lib/common/tango/core/Thread.d	(working copy)
@@ -831,7 +831,8 @@
             {
                 if( !set )
                 {
-                    foreach( Thread t; sm_tbeg )
+                    // foreach( Thread t; sm_tbeg ) Bug in GDC 0.24 SVN
+                    for( Thread t = sm_tbeg; t; t = t.next )
                     {
                         t.m_local[key] = null;
                     }
@@ -858,7 +859,8 @@
         synchronized( slock )
         {
             sm_local[key] = false;
-            foreach( Thread t; sm_tbeg )
+            // foreach( Thread t; sm_tbeg ) Bug in GDC 0.24 SVN
+            for( Thread t = sm_tbeg; t; t = t.next )
             {
                 t.m_local[key] = null;
             }
Index: lib/compiler/gdc/genobj.d
===================================================================
--- lib/compiler/gdc/genobj.d	(revision 2167)
+++ lib/compiler/gdc/genobj.d	(working copy)
@@ -156,7 +156,6 @@
     //  8:                      // has constructors
     void*       deallocator;
     OffsetTypeInfo[] offTi;
-/+
     void function(Object) defaultConstructor;   // default Constructor

     /**
@@ -192,7 +191,6 @@
         }
         return o;
     }
-+/
 }

 /**
